closed/src/java.base/share/native/libjli/criuhelpers.c

```diff
@@ -0,0 +1,309 @@
+ * @param[in] logFile The log file option for CRIU
+ * @return 0 if the execution of the 'criu restore' command succeeds, -1 otherwise
+ */
+static int
+restoreFromCheckpoint(const char *checkpointDirectory, int logLevel, jboolean unprivilegedModeOn, const char *logFile)
+{
+    int restoreStatus = 0;
+    int length = -1;
+    char logLevelOption[4] = { 0 };
+    char *logFileOption = NULL;
+    int argc = 0;
+    const char *argv[9] = { NULL };
+    argv[argc++] = "criu";
+    argv[argc++] = "restore";
+    argv[argc++] = "-D";
+    argv[argc++] = checkpointDirectory;
+    length = snprintf(NULL, 0, "-v%d", logLevel);
+    if (length < 0) {
+        JLI_ReportErrorMessage("Failed to calculate the length of option '-v%d'.", logLevel);
+        restoreStatus = -1;
+        goto done;
+    }
+    if (snprintf(logLevelOption, sizeof(logLevelOption), "-v%d", logLevel) < 0) {
+        JLI_ReportErrorMessage("Failed to format the log level option '-v%d'.", logLevel);
+        restoreStatus = -1;
+        goto done;
+    }
+    argv[argc++] = logLevelOption;
+    argv[argc++] = "--shell-job";
+    if (unprivilegedModeOn) {
+        argv[argc++] = "--unprivileged";
+    }
+    if (NULL != logFile) {
+        length = strlen(logFile) + sizeof("--log-file=%s") - 1;
+        logFileOption = (char *)JLI_MemAlloc(length - 1);
+        if (NULL == logFileOption) {
+            JLI_ReportErrorMessage("Failed to allocate memory for option '--log-file=%s'.", logFile);
+            restoreStatus = -1;
+            goto done;
+        }
+        if (snprintf(logFileOption, length - 1, "--log-file=%s", logFile) < 0) {
+            JLI_ReportErrorMessage("Failed to format option '--log-file=%s'.", logFile);
+            restoreStatus = -1;
+            goto done;
+        }
+        argv[argc++] = logFileOption;
+    }
+    argv[argc] = NULL;
+    execvp(argv[0], (char * const *)argv);
+    /* If execvp returns, there was an error. */
+    restoreStatus = -1;
+freeLogFileOption:
+    if (NULL != logFileOption) {
+        JLI_MemFree((void *)logFileOption);
+        logFileOption = NULL;
+    }
+done:
+    return restoreStatus;
+}
+
+/**
+ * Handle the restoration of the system state from a checkpoint.
+ * @param[in] argc The number of command line arguments
+ * @param[in] argv The array of command line arguments
+ */
+void
+handleCRaCRestore(int argc, char **argv)
+{
+    const char *checkpointDirectory = NULL;
+    int error = 0;
+    int parentProcessExitStatus = EXIT_SUCCESS;
+    pid_t childProcessPid = 0;
+    int logLevel = 0;
+    int childProcessExitStatus = EXIT_SUCCESS;
+    jboolean unprivilegedModeOn = JNI_FALSE;
+    const char *logFile = NULL;
+    int childProcessPidStatus = 0;
+    int childProcessPidExitStatus = 0;
+    checkpointDirectory = getCheckpointDirectory(argc, argv, &error);
+    /* We ignore restore if the restore option -XX:CRaCRestoreFrom is not specified. */
+    if (-1 == error) {
+        return;
```

Comment:
Suggest the comment should be placed where it applies:
```c
    if (-1 == error) {
        /* Option -XX:CRaCRestoreFrom not specified. */
        return;
```

Reply:
Moved the comment.